src/rideables/TGraph.hpp

```diff
@@ -161,171 +258,240 @@ class TGraph : public RGraph{
+                Relation r(src, dest, -1);
+                remove_relation(source(src), &r);
+                remove_relation(destination(dest), &r);
+                inc_seq(src);
+                inc_seq(dest);
             }
 
             if (src > dest) {
-                v1->unlock();
-                v2->unlock();
+                unlock(src);
+                unlock(dest);
             } else {
-                v2->unlock();
-                v1->unlock();
+                unlock(dest);
+                unlock(src);
             }
             return true;
         }
 
-        /**
-         * Sets the label for a node to a specific value
-         * @param id The id the node whose weight to set
-         * @param l The new label for the node
-         */
-        bool set_lbl(int id, int l) {
-            Vertex *v = idxToVertex[id]; 
-            v->lock();
-            v->set_lbl(l);
-            v->unlock();
-            return true;
-        }
+        bool add_vertex(int vid) {
+            std::mt19937_64 vertexGen(time(NULL));
+            std::uniform_int_distribution<> uniformVertex(0,numVertices);
+            bool retval = true;
+            // Randomly sample vertices...
+            std::vector<int> vec(meanEdgesPerVertex);
+            for (size_t i = 0; i < meanEdgesPerVertex * 100 / vertexLoad; i++) {
+                int u = uniformVertex(vertexGen);
+                while (u == i) {
+                    u = uniformVertex(vertexGen);
+                }
+                vec.push_back(u);
+            }
+            vec.push_back(vid);
+            std::sort(vec.begin(), vec.end()); 
+            vec.erase(std::unique(vec.begin(), vec.end()), vec.end());
 
-        /**
-         * Sets the weight for an edge to a specific value. If the edge does not exist, this does not break, but does
-         * unnecessary computation.
-         * @param src the integer id of the source of the edge to set the weight for
-         * @param dest the integer id of the dest of the edge to set the weight for
-         * @param w the new weight value
-         */
-        bool set_weight(int src, int dest, int w) {
-            bool retval = false;
-            // Unimplemented because MontageGraph can't 
+            for (int u : vec) {
+                lock(u);
+            }
+
+            if (idxToVertex[vid] == nullptr) {
+                idxToVertex[vid] = new Vertex(vid, vid);
+                for (int u : vec) {
+                    if (idxToVertex[u] == nullptr) continue;
+                    if (u == vid) continue;
+                    Relation *in = new Relation(vid, u, -1);
+                    Relation *out = new Relation(vid, u, -1);
+                    source(vid).insert(in);
+                    destination(u).insert(out);
+                }
+            } else {
+                retval = false;
+            }
+
+            std::reverse(vec.begin(), vec.end());
```

Comment:
I think `std::reverse` is O(n), and is inside locks. Can we just do `for(auto u=vec.rbegin();u!=vec.rend();u++)` rather than reversing it and then iterating.
Another way, if you want to try new things, C++20 provides `views::reverse`. https://en.cppreference.com/w/cpp/ranges/reverse_view

Reply:
Great point! Makes sense there is a reverse iterator built-in that would be more efficient. Also aren't we compiling for C++17?